atc/creds/vault/vault.go

```diff
@@ -59,10 +65,31 @@ func (v Vault) Get(varDef template.VariableDefinition) (interface{}, bool, error
 	return evenLessTyped, true, nil
 }
 
+func (v Vault) getSecretReader() (SecretReader, error) {
+	v.once.Do(func() {
+		select {
+		case <-v.LoggedIn:
+		case <-time.After(5 * time.Second):
+			v.timeout = true
```

Comment:
`v` is a non-pointer here so I think this assignment won't outlive this method call. That's fine for the first call but later calls won't run the `v.once` and so `v.timeout` will be false. Should this be changed to a pointer receiver?

Reply:
Turns out timeout was only used in this method. So kept `v` as non-pointer and converted `timeout` into a local variable instead

Reply:
Hmm, that makes the assignment clearer but it'll still have the behavior of only attempting to wait on the log-in one time, and any subsequent calls will just start returning the secret reader even though it hasn't logged in yet (since `timeout` will always be `false`).

I guess I'm not clear on the flow here. There's also the fact that there's a new variables object constructed pretty often, so the `.once` is still gonna be evaluated each time. Maybe it should just be removed.

Reply:
@vito here is the summary how I see it:
* getSecretReader() gets called every time a secret is retrieved
* when is gets called for the first time, we'd like to ensure that connection to vault has been established successfully, and give it up to 5 seconds
* if connection has been established successfully, subsequent calls should immediately return the corresponding initialized SecretReader
* if connection has not been established, subsequent calls should return immediately as well (it will be an un-authorized client / SecretReader and all secrets lookup will fail, which seems ok)

The main reason why sync.Once got added was not to avoid creation of time.After, but rather to account for the fact that select (without default clause) is blocking. Without sync.Once, select will read the status off v.LoggedIn for the first time, and then subsequent calls to select would block and wait once again for the timeout duration 

It's probably not obvious from the code, but I'm open to either adding more comments or possibly changing the implementation to make the code more readable.

Reply:
@ralekseenkov Ahh, now I understand the intent. `LoggedIn` is actually *closed* on the first successful log-in, though, not written to on each log-in:

https://github.com/concourse/concourse/blob/996b8d1671411d6d6a80558366bb98958b338f00/atc/creds/vault/reauther.go#L98-L100

So it shouldn't block, because closed channels always return their zero-value.

I think without the `Once` it'll wait up to 5 seconds on each call until the first successful log-in, and at that point it'll stop waiting as the `select` should just immediately unblock. Does that make sense? (Kinda hard to put this into words.)